SpatialGDKEditorToolbar/Source/Private/SchemaGenerator/SchemaGenerator.cpp

```diff
@@ -298,5 +446,210 @@ int GenerateTypeBindingSchema(FCodeWriter& Writer, int ComponentId, UClass* Clas
+		// Handover (server to server) replicated properties.
+		Writer.Printf("component {0} {", *(PropertyName + TEXT("Handover")));
+		Writer.Indent();
+		Writer.Printf("id = {0};", IdGenerator.GetNextAvailableId());
+		Writer.Printf("data {0};", *SchemaHandoverDataName(ComponentClass));
+		Writer.Outdent().Print("}");
+
+		SubobjectData.SchemaComponents[ESchemaComponentType::TYPE_Handover] = IdGenerator.GetCurrentId();
+	}
+
+	FUnrealRPCsByType RPCsByType = GetAllRPCsByType(TypeInfo);
+
+	for (auto Group : GetRPCTypes())
+	{
+		if (RPCsByType[Group].Num() == 0)
+		{
+			continue;
+		}
+
+		Writer.PrintNewLine();
+
+		FString ComponentName = PropertyName + GetRPCTypeName(Group) + TEXT("RPCs");
+		Writer.Printf("component {0} {", *ComponentName);
+		Writer.Indent();
+		Writer.Printf("id = {0};", IdGenerator.GetNextAvailableId());
+		for (auto& RPC : RPCsByType[Group])
+		{
+			if (Group == ERPCType::RPC_NetMulticast)
+			{
+				Writer.Printf("event UnrealRPCCommandRequest {0};",
+					*SchemaRPCName(Cast<UClass>(ComponentClass), RPC->Function));
+			}
+			else
+			{
+				Writer.Printf("command UnrealRPCCommandResponse {0}(UnrealRPCCommandRequest);",
+					*SchemaRPCName(Cast<UClass>(ComponentClass), RPC->Function));
+			}
+		}
+		Writer.Outdent().Print("}");
+
+		if (Group == RPC_Client)
+		{
+			SubobjectData.SchemaComponents[ESchemaComponentType::TYPE_ClientRPC] = IdGenerator.GetCurrentId();
+		}
+		else if (Group == RPC_Server)
+		{
+			SubobjectData.SchemaComponents[ESchemaComponentType::TYPE_ServerRPC] = IdGenerator.GetCurrentId();
+		}
+		else if (Group == RPC_NetMulticast)
+		{
+			SubobjectData.SchemaComponents[ESchemaComponentType::TYPE_NetMulticastRPC] = IdGenerator.GetCurrentId();
+		}
+		else if (Group == RPC_CrossServer)
+		{
+			SubobjectData.SchemaComponents[ESchemaComponentType::TYPE_CrossServerRPC] = IdGenerator.GetCurrentId();
+		}
+	}
+
+	return SubobjectData;
+}
+
+void GenerateSubobjectSchemaForActor(FComponentIdGenerator& IdGenerator, UClass* ActorClass, TSharedPtr<FUnrealType> TypeInfo, FString SchemaPath, FSchemaData& ActorSchemaData)
+{
+	FCodeWriter Writer;
+
+	Writer.Printf(R"""(
+		// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+		// Note that this file has been generated automatically
+		package unreal.generated.{0}.components;)""",
+		*UnrealNameToSchemaTypeName(TypeInfo->Type->GetName().ToLower()));
+
+	Writer.PrintNewLine();
+
+	GenerateActorIncludes(Writer, TypeInfo);
+
+	bool bHasComponents = false;
+	TSet<UObject*> SeenComponents;
+	int32 CurrentOffset = 1;
+
+	for (auto& PropertyPair : TypeInfo->Properties)
+	{
+		UProperty* Property = PropertyPair.Key;
+		UObjectProperty* ObjectProperty = Cast<UObjectProperty>(Property);
+
+		TSharedPtr<FUnrealType>& PropertyTypeInfo = PropertyPair.Value->Type;
+
+		if (ObjectProperty && PropertyTypeInfo.IsValid())
+		{
+			UObject* Value = PropertyTypeInfo->Object;
+
+			if (Value != nullptr && !Value->IsEditorOnly())
+			{
+				if (!SeenComponents.Contains(Value))
```

Comment:
Under which circumstances does an object get here multiple times?

Reply:
A root component can be the same as a variable that points to a component, for example. Any component can have multiple references to it.